contracts/periphery/astETH.sol

```diff
@@ -1,34 +1,79 @@
 import {ERC20} from "@openzeppelin/contracts/token/ERC20/ERC20.sol";
 import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
 import {SafeERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import {Ownable2Step} from "@openzeppelin/contracts/access/Ownable2Step.sol";
 
-contract astETH is ERC20 {
-    using SafeERC20 for IERC20;
+interface IStETH is IERC20 {
+    function submit(address _referral) external payable returns (uint256);
+}
+
+interface IAstETH {
+    event FeeRecipientSet(address feeRecipient);
+    event FeeCollected(address feeRecipient, uint256 feeCollected);
+}
 
-    address public immutable stETH;
-    address public immutable feeRecipient;
+contract AstETH is IAstETH, ERC20, Ownable2Step {
+    using SafeERC20 for IStETH;
+
+    IStETH public immutable stETH;
+    address public feeRecipient;
 
     constructor(
-        address _stETH,
+        IStETH _stETH,
+        address _owner,
         address _feeRecipient
     ) ERC20("Alongside stETH", "astETH") {
         stETH = _stETH;
         feeRecipient = _feeRecipient;
+        _transferOwnership(_owner);
     }
 
-    function deposit(uint256 amount) external {
-        IERC20(stETH).safeTransferFrom(msg.sender, address(this), amount);
-        _mint(msg.sender, amount);
+    ///////////////////////// ISSUANCE /////////////////////////
+
+    /// @notice Deposit stETH to receive astETH
+    /// @param  amountToMint stETH amount to deposit
+    function deposit(uint256 amountToMint) external returns (uint256) {
+        stETH.safeTransferFrom(msg.sender, address(this), amountToMint);
+        _mint(msg.sender, amountToMint);
+        return amountToMint;
     }
 
-    function withdraw(uint256 amount) external {
-        _burn(msg.sender, amount);
-        IERC20(stETH).safeTransfer(msg.sender, amount);
+    /// @notice Wraps ETH to stETH before depositing stETH to receive astETH
+    function wrapAndDeposit() external payable returns (uint256) {
+        uint256 balanceBefore = stETH.balanceOf(address(this));
+        stETH.submit{value: msg.value}(address(0));
```

Comment:
Note to self: I want to make sure there's no reentrancy risk here. If there is, could reenter and call `wrapAndDeposit()` again, and then we'd get double increment on the outer loop balance.

Reply:
Confirmed, no callbacks in this function. You could add reentrancy check if you want to be extra paranoid.